smithy-build/src/main/java/software/amazon/smithy/build/transforms/RemoveTraitDefinitions.java

```diff
@@ -0,0 +1,106 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.build.transforms;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.logging.Logger;
+import software.amazon.smithy.build.TransformContext;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.TraitDefinition;
+import software.amazon.smithy.model.transform.ModelTransformer;
+
+/**
+ * {@code removeTraitShapes} removes all trait definitions from a model,
+ * but leaves if the trait definition contains any of the provided
+ * {@code tags}.
+ *
+ * <p>This transformer will not remove prelude trait definitions.
+ */
+public final class RemoveTraitDefinitions extends BackwardCompatHelper<RemoveTraitDefinitions.Config> {
+
+    private static final Logger LOGGER = Logger.getLogger(RemoveTraitDefinitions.class.getName());
+
+    /**
+     * {@code removeTraitShapes} configuration settings.
+     */
+    public static final class Config {
+
+        private Set<String> exportTagged = Collections.emptySet();
+
+        /**
+         * You can <em>export</em> trait definitions by applying specific tags
+         * to the trait definition and adding the list of export tags as an
+         * argument to the transformer.
+         *
+         * @param exportByTags Tags that cause trait definitions to be
+         *                     exported.
+         */
+        public void setExportTagged(Set<String> exportByTags) {
+            this.exportTagged = exportByTags;
+        }
+
+        /**
+         * Gets the set of tags that are used to export trait definitions.
+         *
+         * @return the tags that are used to export trait definitions.
+         */
+        public Set<String> getExportTagged() {
+            return exportTagged;
+        }
+    }
+
+    @Override
+    String getBackwardCompatibleNameMapping() {
+        return "removeTraitDefinitions";
+    }
+
+    @Override
+    public Class<Config> getConfigType() {
+        return Config.class;
+    }
+
+    @Override
+    protected Model transformWithConfig(TransformContext context, Config config) {
+        Model model = context.getModel();
+        ModelTransformer transformer = context.getTransformer();
+
+        Model modelWithoutTraitDefinitions = transformer.getModelWithoutTraitShapes(model);
+
+        // If no export tags have been configured, we can return the scrubbed model directly.
+        if (config.getExportTagged().isEmpty()) {
+            return modelWithoutTraitDefinitions;
+        }
+
+        Predicate<Shape> keepTraitDefsByTag = trait -> config.getExportTagged().stream().anyMatch(trait::hasTag);
+        Model.Builder builder = modelWithoutTraitDefinitions.toBuilder();
+
+        // Add trait definitions back to scrubbed model.
+        model.shapes()
```

Comment:
I don't see how this would work if, say, a trait shape is removed that is a structure that references other shapes that were removed. It might be better to instead filter before removing trait shapes rather than re-add. (e.g., maybe you can add a `Predicate` to the `getModelWithoutTraitShapes` method.